base/sparse/sparsematrix.jl

```diff
@@ -1884,48 +1884,61 @@ function setindex!{Tv,Ti,T<:Real}(A::SparseMatrixCSC{Tv,Ti}, x, I::AbstractVecto
 end
 
 
-
 # Sparse concatenation
 
 function vcat(X::SparseMatrixCSC...)
+    Tv = promote_type(map(x->eltype(x.nzval), X)...)
+    Ti = promote_type(map(x->eltype(x.rowval), X)...)
+
+    vcat(map(x->convert(SparseMatrixCSC{Tv,Ti}, x), X)...)
```

Comment:
Rather than effectively converting the input arrays twice, another approach might be to break out the "stuff a single matrix into its slot" algorithm into a subfunction. That way the performance-intensive part will be type-stable.